src/items/recipes.zig

```diff
@@ -0,0 +1,226 @@
+					return error.NoMatch;
+				}
+				idx += literal.len;
+			},
+			.symbol => |symbol| {
+				var indexes: main.List(usize) = .init(allocator);
+				defer indexes.deinit();
+				if(i + 1 < pattern.len) {
+					const nextSegment = pattern[i + 1];
+					var nextIndex = idx;
+					while(std.mem.indexOfPos(u8, target, nextIndex, nextSegment.literal)) |endIndex| {
+						indexes.append(endIndex);
+						nextIndex = endIndex + 1;
+					}
+				} else {
+					indexes.append(target.len);
+				}
+				if(indexes.items.len == 0) {
+					return error.NoMatch;
+				}
+				if(indexes.items.len == 1) {
+					if(newKeys.get(symbol)) |value| {
+						if(!std.mem.eql(u8, target[idx..indexes.items[0]], value)) {
+							return error.NoMatch;
+						}
+					} else {
+						newKeys.put(allocator.dupe(u8, symbol), allocator.dupe(u8, target[idx..indexes.items[0]])) catch unreachable;
+					}
+					idx = indexes.items[0];
+				} else {
+					defer newKeys.deinit();
+					var newKeyPairs: main.List(std.StringHashMap([]const u8)) = .init(allocator);
+					defer newKeyPairs.deinit();
+					for(indexes.items) |endIndex| {
+						if(matchWithKeys(allocator, target[endIndex..], pattern[i + 1 ..], &newKeys)) |newKeyMatches| {
+							newKeyPairs.appendSlice(newKeyMatches);
+							allocator.free(newKeyMatches);
+						} else |_| {}
+					}
+					if(newKeyPairs.items.len == 0) return error.NoMatch;
+					return newKeyPairs.toOwnedSlice();
+				}
+			},
+		}
+	}
+	if(idx == target.len) {
+		var newKeyPairs = allocator.alloc(std.StringHashMap([]const u8), 1);
+		newKeyPairs[0] = newKeys;
+		return newKeyPairs;
+	} else {
+		return error.NoMatch;
+	}
+}
+
+const ItemWithAmount = struct {
+	item: BaseItemIndex,
+	amount: u16,
+};
+
+const ItemKeyPair = struct {item: ItemWithAmount, keys: std.StringHashMap([]const u8)};
+
+fn findRecipeItemOptions(allocator: NeverFailingAllocator, itemStackPattern: ItemStackPattern, keys: *const std.StringHashMap([]const u8)) ![]ItemKeyPair {
+	const pattern = itemStackPattern.pattern;
+	const amount = itemStackPattern.amount;
+
+	if(pattern.len == 1 and pattern[0] == .literal) {
+		const item = BaseItemIndex.fromId(pattern[0].literal) orelse return error.ItemNotFound;
+		return allocator.dupe(ItemKeyPair, &.{.{
+			.item = .{
+				.item = item,
+				.amount = amount,
+			},
+			.keys = keys.clone() catch unreachable,
+		}});
+	}
+	var itemPairs: main.List(ItemKeyPair) = .initCapacity(allocator, 1);
+	defer itemPairs.deinit();
+	var iter = items.iterator();
+	while(iter.next()) |item| {
+		const newKeyMatches = matchWithKeys(allocator, item.id(), pattern, keys) catch continue;
+		for(newKeyMatches) |match| {
+			itemPairs.append(.{
+				.item = .{
+					.item = item.*,
+					.amount = amount,
+				},
+				.keys = match,
+			});
+		}
+	}
+	return itemPairs.toOwnedSlice();
+}
+
+fn generateItemCombos(allocator: NeverFailingAllocator, recipe: []ZonElement) ![][]ItemWithAmount {
```

Comment:
Please do a bit more const correctness
```suggestion
fn generateItemCombos(allocator: NeverFailingAllocator, recipe: []const ZonElement) ![]const []const ItemWithAmount {
```